### PR TITLE
Port Indirect Corrections Interface to Workbench

### LINKED ISSUE #24373: Port Indirect Corrections GUI to workbench
This will be the first example of porting a C++ interface to the new workbench.

Plan
----

* Define `InterfacePlugin` class in C++ and export this to Python with sip in the `mantidqt` package.
* Create a Python-based `InterfaceFactory` implementation in `mantidqt`
* Define a sip wrapper for each scientific interface library and make the constructor available in Python
* Create a `scientific_interfaces` directory in the installation in the `lib` directory
  * It would contain, for example, an `Indirect` directory:
```
 - lib/scientific_interfaces/IndirectGui/
                                     __init__.py
                                     _indirectqt5.so
```
* Create new `workbench/plugins/interfaces` directory and
  * for example for the indirect corrections interface create a file called `register_indirect.py` in the above directory with content such as
  ```python
  from mantidqt.plugins import InterfacePlugin, InterfaceFactory
  
  from IndirectGui import IndirectCorrections
  
  class CorrectionsInterface(InterfacePlugin):
      def name():
          return "Corrections"
          
      def category():
          # Controls the menu placement
          return "Indirect"
  
      def create(parent=None, show=False)
         window = IndirectCorrections(parent)
         if show:
             window.show()
         return window
  
  InterfaceFactory.subscribe(CorrectionsInterface())
  ```
  
  The workbench can then use the `InterfaceFactory` to create a named interface and show it. We can do something similar with the Python GUIs
  and move them out of `scripts`.
  
Steps
-----
1. Make indirect interface library Qt5 compatible (requires replacing the Qwt plotting with the matplotlib C++ library)
1. Implement `InterfacePlugin` class in C++ common widgets
1. Implement `InterfaceFactory` in Python
1. Create thin sip wrapper around Indirect library
1. Tie it all together in the workbench

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work.**

Creates a Qt5 build of `qt/scientific_interfaces/Indirect.` This PR only contains the `Corrections` interface and the changes required to use a matplotlib-based canvas embedded in the interface.

Note that this does not implement all C++-based plotting required by the interface. Some functionality is coming in other pull requests: #25712 and #25617 that will make this significantly easier to achieve. I feel that it is better to get this bulk reviewed and in before adding the other functionality. The things that are missing:
* the error bars option on the context menu: #25617 enables this on standard plots and it will be easier to just reuse this functionality for the embedded plots rather than reimplementing it
* the preview plot button does nothing as it creates a plot in a separate window and this requires #25712

**To test:**

Start the Corrections interface in both MantidPlot and Workbench and test the functionality.

Refs #24373

*This does not require release notes* because **the release notes for indirect will be updated when more of their useful interfaces are ported.**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by robertapplin** [comment]
When testing on MantidPlot I get an error:
1. `Corrections`->`Absorption Corrections`
2. Load the files below
3. Click `Run`, and error!

**Data Files**
Sample: [irs26176_graphite002_red.zip](https://github.com/mantidproject/mantid/files/3187600/irs26176_graphite002_red.zip)
Container: [irs26174_graphite002_red.zip](https://github.com/mantidproject/mantid/files/3187602/irs26174_graphite002_red.zip)

**Comment by robertapplin** [comment]
Same error occurs on MantidPlot when:
1. `Corrections`->`CalculatePaalmanPings`
2. Change the `Mass Density` combo box to say `Atom Number Density`

This one may be quicker to replicate.

File: qt/widgets/mplcpp/src/PanZoomTool.cpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+constexpr auto TOOLBAR_MODE_ATTR = "mode";
+constexpr auto TOOLBAR_ZOOM_MODE_STR = "zoom rect";
+constexpr auto TOOLBAR_ZOOM_METHOD = "zoom";
+constexpr auto TOOLBAR_PAN_MODE_STR = "pan/zoom";
+constexpr auto TOOLBAR_PAN_METHOD = "pan";
+
+/// Return the matplotlib NavigationToolbar type appropriate
+/// for our backend. It is returned hidden
+Python::Object mplNavigationToolbar(FigureCanvasQt *canvas) {
+  auto backend = backendModule();
```
Comment (robertapplin) [inline qt/widgets/mplcpp/src/PanZoomTool.cpp]: `const auto backend`?

File: qt/widgets/mplcpp/test/FigureTest.h [hunk 2]
```diff
@@ -43,20 +43,36 @@ class FigureTest : public CxxTest::TestSuite {
...
 
-  void testSetFaceColor() {
+  void testSetFaceColorFromString() {
     Figure fig{false};
     fig.setFaceColor("r");
-    auto rgb = fig.pyobj().attr("get_facecolor")();
-    TS_ASSERT_EQUALS(1.0, rgb[0]);
-    TS_ASSERT_EQUALS(0.0, rgb[1]);
-    TS_ASSERT_EQUALS(0.0, rgb[2]);
+    auto color = fig.faceColor();
```
Comment (robertapplin) [inline qt/widgets/mplcpp/test/FigureTest.h]: Is `const auto color` possible?

**Comment by martyngigg** [comment]
Thanks, @robertapplin. I had been so focussed on the plot I had not hit the `Run` button :man_facepalming: They all look like the same cause and that should be fixed now.

**Review by robertapplin — CHANGES_REQUESTED** [review]
The code changes look very good, I have a few comments below which may or may not be possible. I'll continue with the interface testing now.

File: qt/widgets/mplcpp/inc/MantidQtWidgets/MplCpp/Figure.h [hunk 3]
```diff
@@ -42,9 +42,11 @@ class MANTID_MPLCPP_DLL Figure : public Python::InstanceHolder {
     return Axes{pyobj().attr("axes")[index]};
   }
 
+  QColor faceColor() const;
+  void setFaceColor(QColor color);
   void setFaceColor(const char *color);
   Axes addAxes(double left, double bottom, double width, double height);
-  Axes addSubPlot(int subplotspec);
+  Axes addSubPlot(const int subplotspec, QString projection = "");
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply qt/widgets/mplcpp/inc/MantidQtWidgets/MplCpp/Figure.h]: Yes absolutely. Good catch.

File: qt/widgets/mplcpp/inc/MantidQtWidgets/MplCpp/FigureCanvasQt.h [hunk 4]
```diff
@@ -27,15 +27,19 @@ class MANTID_MPLCPP_DLL FigureCanvasQt : public QWidget,
...
   FigureCanvasQt(Figure fig, QWidget *parent = nullptr);
 
   /// Attach an event filter to the underlying matplotlib canvas
   void installEventFilterToMplCanvas(QObject *filter);
   /// Access to the current figure instance.
   inline Figure gcf() const { return m_figure; }
+
   /// Access to the current active axes instance.
-  inline Axes gca() const { return m_figure.gca(); }
+  template <typename AxesType = Axes> inline AxesType gca() const {
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply qt/widgets/mplcpp/inc/MantidQtWidgets/MplCpp/FigureCanvasQt.h]: I was mimicking the matplotlib api here. `gca` stands for "get current axes".

File: qt/widgets/mplcpp/src/PanZoomTool.cpp (hunk 1 above)
Comment (martyngigg) [reply qt/widgets/mplcpp/src/PanZoomTool.cpp]: :+1:

File: qt/widgets/mplcpp/src/Axes.cpp [hunk 5]
```diff
@@ -79,6 +102,47 @@ void Axes::setTitle(const char *label) {
...
+ * @param xdata A vector containing the X data
+ * @param ydata A vector containing the Y data
+ * @param format A format string accepted by matplotlib.axes.Axes.plot.
+ * @param label A label for the line
+ * @return A new Line2D object that owns the xdata, ydata vectors. If the
+ * return value is not captured the line will be automatically removed from
+ * the canvas as the vector data will be destroyed.
+ */
+Line2D Axes::plot(std::vector<double> xdata, std::vector<double> ydata,
+                  QString format, QString label) {
```
Comment (robertapplin) [inline qt/widgets/mplcpp/src/Axes.cpp]: Is `const QString format, const Qstring label` possible?

File: qt/widgets/mplcpp/src/Figure.cpp [hunk 6]
```diff
@@ -68,12 +88,20 @@ Axes Figure::addAxes(double left, double bottom, double width, double height) {
...
+ * @return A wrapper around the Axes object
  */
-Axes Figure::addSubPlot(int subplotspec) {
+Axes Figure::addSubPlot(const int subplotspec, const QString projection) {
   GlobalInterpreterLock lock;
-  return Axes{pyobj().attr("add_subplot")(subplotspec)};
+  if (projection.isEmpty())
+    return Axes{pyobj().attr("add_subplot")(subplotspec)};
+  else {
+    auto args = Python::NewRef(Py_BuildValue("(i)", subplotspec));
```
Comment (robertapplin) [inline qt/widgets/mplcpp/src/Figure.cpp]: `const auto args`?

File: qt/widgets/mplcpp/src/MantidAxes.cpp [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+ * @param lineColour Set the line colour to this string name
+ * @param label A label for the curve
+ * @return A new Line2D artist object
+ */
+Line2D MantidAxes::plot(const Mantid::API::MatrixWorkspace_sptr &workspace,
+                        const size_t wkspIndex, const QString lineColour,
+                        const QString label) {
+  GlobalInterpreterLock lock;
+  auto wksp = Python::NewRef(MatrixWorkpaceToPython()(workspace));
+  auto args = Python::NewRef(Py_BuildValue("(O)", wksp.ptr()));
```
Comment (robertapplin) [inline qt/widgets/mplcpp/src/MantidAxes.cpp]: Is `const auto`'s possible?

File: qt/widgets/mplcpp/test/AxesTest.h [hunk 8]
```diff
@@ -23,6 +23,31 @@ class AxesTest : public CxxTest::TestSuite {
     TS_ASSERT_THROWS_NOTHING(Axes axes(pyAxes()));
   }
 
+  void testClear() {
+    Axes axes(pyAxes());
+    auto line = axes.plot({1, 2, 3}, {1, 2, 3});
```
Comment (robertapplin) [inline qt/widgets/mplcpp/test/AxesTest.h]: Same comment here with the `const`'s, I'm being very picky!

File: dev-docs/source/MatplotlibInCpp.rst [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+Overview
+--------
+
+This document describes the functionality with the C++ layer for interacting with matplotlib plotting.
+
+Embedding Figures
+-----------------
+
+Matplotlib is designed such that figures can be embedded within other Qt widgets and not just created as seperate
```
Comment (robertapplin) [inline dev-docs/source/MatplotlibInCpp.rst]: `separate` instead of `seperate`?

File: dev-docs/source/MatplotlibInCpp.rst [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+     // and when the Line2D is destroyed it will be removed from the canvas.
+     m_line = m_canvas->gca().plot(xpoints, ypoints);
+     // Most operations on the canvas don't automatically redraw the canvas
+     // so drawing must be forced
+     m_canvas->draw();
+   }
+
+The most important part of the above example is capturing the returned ``Line2D`` object.
+In this basic mode with vectors the library avoids copying the data vectors but this means that
+something needs to keep the data alive while the is displayed.
```
Comment (robertapplin) [inline dev-docs/source/MatplotlibInCpp.rst]: I think the last `the` is meant to be `it?

Comment (martyngigg) [reply dev-docs/source/MatplotlibInCpp.rst]: :+1:

File: dev-docs/source/MatplotlibInCpp.rst (hunk 9 above)
Comment (martyngigg) [reply dev-docs/source/MatplotlibInCpp.rst]: :+1:

**Review by robertapplin — APPROVED** [review]
Code looks good and the corrections interface works well 👍 . There are a couple of minor issues that will need fixing in separate PR's (#25828 and #25829). Merging this PR now will allow the other indirect interfaces to be ported sooner.

**Comment by martyngigg** [comment]
@mantidproject/gatekeepers I would appreciate this being merged if possible? Reviewing some other work will be significantly easier with this in master.

**Comment by martyngigg** [comment]
Thanks!

**Pull request merged**